Flight tutorials

benarmston opened the pull request:
@mjtko / @jamesremuscat I'm not sure who best to direct this to.  Would one of you please take a look and let me know if this is heading in the right direction?

Also, I'd appreciate being pointed to instructions on how to test this.

Note: The `sysv` init file is unlikely to work.  I need to figure out how to deal with `PID_FILE` or lack of one.

Thanks.

benarmston committed:
First pass at flight-tutorials

mjtko commented on alces-flight-tutorials/etc/alces-flight-www/tutorial.conf.tpl:
```diff
@@ -0,0 +1,32 @@
+}
+
+location @tutorial {
+    proxy_pass http://127.0.0.1:3000;
```
I think we should avoid port 3000, as it's a "common" port.  Perhaps throw it up in the 25000s somewhere? Say, 25288?

mjtko commented on alces-flight-tutorials/init/sysv/clusterware-alces-flight-tutorials.el6:
```diff
@@ -0,0 +1,120 @@
+  echo -n Starting Alces Flight Tutorials Server:
+  cd $DAEMON_HOME
+  $DAEMON_CMD
+  sleep 2
```
I would imagine that inserting this the `sleep` would be sufficient to generate the PID file:

```
echo "$!" > "${PID_FILE}"
```

Don't worry too much about the sysvinit script, but it's good to have a reasonable effort for completeness!

mjtko commented on alces-flight-tutorials/metadata.yml:
```diff
@@ -0,0 +1,75 @@
+  el6: |
+    yum install -y -e0 nodejs npm yarn
+  el7: |
+    yum install -y -e0 nodejs npm yarn
```
I suspect none of these packages are available in EL7 and you'll need to enable EPEL.  e.g. https://github.com/alces-software/clusterware-sessions/blob/master/xfce/metadata.yml#L14-L21

Looks like there's no `yarn` package in EPEL.  I guess that's why you need `npm`, so `yarn` can be installed?  Is `yarn` required after build or can we use `npm run` or something instead?

mjtko commented on alces-flight-tutorials/metadata.yml:
```diff
@@ -0,0 +1,75 @@
+    if serviceware_is_installed alces-flight-www; then
+        cp etc/alces-flight-www/tutorial.conf.tpl \
+          "${cw_ROOT}"/etc/alces-flight-www/server-https.d/tutorial.conf
+        # XXX Does this work with relative paths?
```
I don't remember, but I suspect not.  You might get some mileage out of:

`$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)` to get the absolute path.

mjtko commented on alces-flight-tutorials/metadata.yml:
```diff
@@ -0,0 +1,75 @@
+
+build:
+  el6: |
+    yum install -y -e0 nodejs yarn
```
EPEL required here too I reckon.

mjtko commented on alces-flight-tutorials/metadata.yml:
```diff
@@ -0,0 +1,75 @@
+    tar -C /tmp/flight-tutorials-client --strip-components=1 -xzf /tmp/flight-tutorials-client.tar.gz
+
+    pushd /tmp/flight-tutorials-client
+    yarn run build
```
Think you might need to install `yarn` via `npm` first.

mjtko commented on alces-flight-tutorials/resource/alces-flight-www/tutorial/config.yml:
```diff
@@ -0,0 +1,9 @@
+short_title: Tutorial
+paras:
+ - >-
+   Get up and running with this simple Pegamoose-infused Alces Flight
```
This copy could do with being updated. 😄

mjtko reviewed (commented):
On the right track. Few comments inline.

For testing, you can spin up a Vagrant VM, or go the whole hog and build an AMI.

Vagrant VM: @bobwhitelock wrote some notes here: https://github.com/alces-software/imageware/blob/master/support/vagrant/DEVELOPMENT.md
AMI: you'll need to create a branch of `clusterware` which changes `scripts/registry.rc` to point at the branch of `clusterware-services` you want to test. Then build as usual using the `ami-creator` tool here: https://github.com/alces-software/imageware/tree/develop/support/aws

I'd start with the VM approach and graduate to the AMI approach once you're more confident that the build is correct.

Note that, until you've built and uploaded a tarball to the S3 bucket that contains the clusterware service tarballs, you'll need to use `alces service build` rather than `alces service install`.

benarmston committed:
Use a higher (less commonly used) port

benarmston committed:
Populate PID_FILE

benarmston committed:
Enable EPEL; use npm not yarn

benarmston committed:
Replace relative path with absolute path

benarmston committed:
Update some copy

benarmston committed:
Install dependencies before building

benarmston committed:
Build into ${cw_ROOT}/opt/ not /opt/

benarmston committed:
Remove any previous build before building again

benarmston committed:
Remove use of `local` outside of function

benarmston committed:
The install script now actually you know installs

benarmston committed:
Fix issues with systemd service file

 - Use absolute paths and /bin/env
 - Add ExecStop command
 - Use username not UID

benarmston committed:
Fix working dir

benarmston committed:
Naming consistency tutorial -> alces-flight-tutorials

Except for the URL which has tutorial -> tutorials

benarmston committed:
We need git to install flight-common

benarmston committed:
Restart alces-flight-www after enabling flight-tutorials

benarmston committed:
Use git already installed as part of clusterware

benarmston committed:
Let's start the tutorials service when enabled

benarmston commented:
@mjtko a gazillion commits later and I think that this is done and ready for review/merge.
